File: src/cpp_module/gif.cpp
// Python の py_module.gif_helper に対応する GIF 書き出し実装です。

#include "cpp_module/gif.h"

#include <cstdint>
#include <fstream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

namespace pycs::cpp_module::gif {
namespace {

void append_u16_le(std::string& out, std::uint16_t v) {
    out.push_back(static_cast<char>(v & 0xFF));
    out.push_back(static_cast<char>((v >> 8) & 0xFF));
}

std::string lzw_encode(const std::string& data, int min_code_size) {
    if (data.empty()) {
        return std::string();
    }

    const int clear_code = 1 << min_code_size;
    const int end_code = clear_code + 1;

    std::unordered_map<std::string, int> dict;
    dict.reserve(5000);
    for (int i = 0; i < clear_code; ++i) {
        dict[std::string(1, static_cast<char>(i))] = i;
    }

    int next_code = end_code + 1;
    int code_size = min_code_size + 1;

    std::string out;
    out.reserve(data.size());

    std::uint32_t bit_buffer = 0;
    int bit_count = 0;

    auto emit = [&](int code) {
        bit_buffer |= (static_cast<std::uint32_t>(code) << bit_count);
        bit_count += code_size;
        while (bit_count >= 8) {
            out.push_back(static_cast<char>(bit_buffer & 0xFF));
            bit_buffer >>= 8;
            bit_count -= 8;
        }
    };

    emit(clear_code);

    std::string s(1, data[0]);
    for (std::size_t i = 1; i < data.size(); ++i) {
        const char c = data[i];
        std::string sc = s;
        sc.push_back(c);

        auto it = dict.find(sc);
        if (it != dict.end()) {
            s = sc;
            continue;
        }

        emit(dict[s]);

        if (next_code < 4096) {
            dict.emplace(sc, next_code);
            ++next_code;
            if (next_code == (1 << code_size) && code_size < 12) {
                ++code_size;
            }
        } else {
            emit(clear_code);
            dict.clear();
            for (int j = 0; j < clear_code; ++j) {
                dict[std::string(1, static_cast<char>(j))] = j;
            }
            next_code = end_code + 1;
            code_size = min_code_size + 1;
        }

        s = std::string(1, c);
    }

    emit(dict[s]);
    emit(end_code);

    if (bit_count > 0) {
        out.push_back(static_cast<char>(bit_buffer & 0xFF));
    }

    return out;
}

}  // namespace

std::string grayscale_palette() {
    std::string p;
    p.reserve(256 * 3);
    for (int i = 0; i < 256; ++i) {
        const char v = static_cast<char>(i);
        p.push_back(v);
        p.push_back(v);
        p.push_back(v);
    }
    return p;
}

void save_gif(
    const std::string& path,
    int width,
    int height,
    const std::vector<std::string>& frames,
    const std::string& palette,
    int delay_cs,
    int loop
) {
    if (width <= 0 || height <= 0) {
        throw std::runtime_error("gif: width/height must be positive");
    }
    if (palette.size() != 256 * 3) {
        throw std::runtime_error("gif: palette must be 768 bytes");
    }

    const std::size_t frame_bytes = static_cast<std::size_t>(width) * static_cast<std::size_t>(height);
    for (const auto& fr : frames) {
        if (fr.size() != frame_bytes) {
            throw std::runtime_error("gif: frame size mismatch");
        }
    }

    std::string out;
    out.reserve(static_cast<std::size_t>(1024) + frames.size() * frame_bytes / 2);

    out.append("GIF89a", 6);
    append_u16_le(out, static_cast<std::uint16_t>(width));
    append_u16_le(out, static_cast<std::uint16_t>(height));
    out.push_back(static_cast<char>(0xF7));
    out.push_back(static_cast<char>(0));
    out.push_back(static_cast<char>(0));
    out.append(palette);

    out.push_back(static_cast<char>(0x21));
    out.push_back(static_cast<char>(0xFF));
    out.push_back(static_cast<char>(0x0B));
    out.append("NETSCAPE2.0", 11);
    out.push_back(static_cast<char>(0x03));
    out.push_back(static_cast<char>(0x01));
    append_u16_le(out, static_cast<std::uint16_t>(loop));
    out.push_back(static_cast<char>(0x00));

    for (const auto& fr : frames) {
        out.push_back(static_cast<char>(0x21));
        out.push_back(static_cast<char>(0xF9));
        out.push_back(static_cast<char>(0x04));
        out.push_back(static_cast<char>(0x00));
        append_u16_le(out, static_cast<std::uint16_t>(delay_cs));
        out.push_back(static_cast<char>(0x00));
        out.push_back(static_cast<char>(0x00));

        out.push_back(static_cast<char>(0x2C));
        append_u16_le(out, static_cast<std::uint16_t>(0));
        append_u16_le(out, static_cast<std::uint16_t>(0));
        append_u16_le(out, static_cast<std::uint16_t>(width));
        append_u16_le(out, static_cast<std::uint16_t>(height));
        out.push_back(static_cast<char>(0x00));

        out.push_back(static_cast<char>(0x08));
        const std::string compressed = lzw_encode(fr, 8);
        std::size_t pos = 0;
        while (pos < compressed.size()) {
            const std::size_t len = (compressed.size() - pos > 255) ? 255 : (compressed.size() - pos);
            out.push_back(static_cast<char>(len));
            out.append(compressed, pos, len);
            pos += len;
        }
        out.push_back(static_cast<char>(0x00));
    }

    out.push_back(static_cast<char>(0x3B));

    std::ofstream ofs(path, std::ios::binary);
    if (!ofs) {
        throw std::runtime_error("gif: failed to open output file");
    }
    ofs.write(out.data(), static_cast<std::streamsize>(out.size()));
}

}  // namespace pycs::cpp_module::gif

File: src/pytra/utils/gif.py
"""アニメーションGIFを書き出すための最小ヘルパー。"""


from pytra.std import abi


def _gif_append(dst: bytearray, src: bytearray) -> None:
    i = 0
    n = len(src)
    while i < n:
        dst.append(src[i])
        i += 1


def _gif_u16le(v: int) -> bytearray:
    return bytearray([v & 0xFF, (v >> 8) & 0xFF])


def _lzw_encode(data: bytes, min_code_size: int = 8) -> bytes:
    if len(data) == 0:
        empty: bytearray = bytearray()
        return bytes(empty)

    clear_code = 1 << min_code_size
    end_code = clear_code + 1
    code_size = min_code_size + 1

    out: bytearray = bytearray()
    bit_buffer = 0
    bit_count = 0

    bit_buffer |= clear_code << bit_count
    bit_count += code_size
    while bit_count >= 8:
        out.append(bit_buffer & 0xFF)
        bit_buffer = bit_buffer >> 8
        bit_count -= 8
    code_size = min_code_size + 1

    for v in data:
        bit_buffer |= v << bit_count
        bit_count += code_size
        while bit_count >= 8:
            out.append(bit_buffer & 0xFF)
            bit_buffer = bit_buffer >> 8
            bit_count -= 8

        bit_buffer |= clear_code << bit_count
        bit_count += code_size
        while bit_count >= 8:
            out.append(bit_buffer & 0xFF)
            bit_buffer = bit_buffer >> 8
            bit_count -= 8

        code_size = min_code_size + 1

    bit_buffer |= end_code << bit_count
    bit_count += code_size
    while bit_count >= 8:
        out.append(bit_buffer & 0xFF)
        bit_buffer = bit_buffer >> 8
        bit_count -= 8

    if bit_count > 0:
        out.append(bit_buffer & 0xFF)

    return bytes(out)


def grayscale_palette() -> bytes:
    p: bytearray = bytearray()
    i = 0
    while i < 256:
        p.append(i)
        p.append(i)
        p.append(i)
        i += 1
    return bytes(p)


@abi(args={"frames": "value"})
def save_gif(
    path: str,
    width: int,
    height: int,
    frames: list[bytes],
    palette: bytes,
    delay_cs: int = 4,
    loop: int = 0,
) -> None:
    if len(palette) != 256 * 3:
        raise ValueError("palette must be 256*3 bytes")

    frame_lists: list[bytearray] = []
    for fr in frames:
        fr_buf: bytearray = bytearray()
        for v in fr:
            fr_buf.append(int(v))
        if len(fr_buf) != width * height:
            raise ValueError("frame size mismatch")
        frame_lists.append(fr_buf)

    palette_buf: bytearray = bytearray()
    for v in palette:
        palette_buf.append(int(v))

    out: bytearray = bytearray()
    _gif_append(out, bytearray([71, 73, 70, 56, 57, 97]))  # GIF89a
    _gif_append(out, _gif_u16le(width))
    _gif_append(out, _gif_u16le(height))
    out.append(0xF7)
    out.append(0)
    out.append(0)
    _gif_append(out, palette_buf)

    _gif_append(out, bytearray([0x21, 0xFF, 0x0B, 78, 69, 84, 83, 67, 65, 80, 69, 50, 46, 48, 0x03, 0x01]))
    _gif_append(out, _gif_u16le(loop))
    out.append(0)

    for fr_buf in frame_lists:
        _gif_append(out, bytearray([0x21, 0xF9, 0x04, 0x00]))
        _gif_append(out, _gif_u16le(delay_cs))
        _gif_append(out, bytearray([0x00, 0x00]))

        out.append(0x2C)
        _gif_append(out, _gif_u16le(0))
        _gif_append(out, _gif_u16le(0))
        _gif_append(out, _gif_u16le(width))
        _gif_append(out, _gif_u16le(height))
        out.append(0)
        out.append(8)
        compressed = _lzw_encode(bytes(fr_buf), 8)
        pos = 0
        while pos < len(compressed):
            remain = len(compressed) - pos
            chunk_len = 255 if remain > 255 else remain
            out.append(chunk_len)
            i = 0
            while i < chunk_len:
                out.append(compressed[pos + i])
                i += 1
            pos += chunk_len
        out.append(0)

    out.append(0x3B)

    with open(path, "wb") as f:
        f.write(bytes(out))
